src/filter/bilateral.rs

```diff
@@ -0,0 +1,288 @@
+        debug_assert!(image.in_bounds(x, y));
+        // Safety: `Image::from_fn` yields `col` in [0, width) and `row` in [0, height).
+        let center_pixel = unsafe { image.unsafe_get_pixel(x, y) };
+
+        let window_len = 2 * radius + 1;
+        let weights_and_values = window_range
+            .clone()
+            .cartesian_product(window_range.clone())
+            .map(|(w_y, w_x)| {
+                // these casts will always be correct due to asserts made at the beginning of the
+                // function about the image width/height
+                //
+                // the subtraction will also never overflow due to the `is_empty()` assert
+                let window_y = (i32::from(w_y) + (y as i32)).clamp(0, (height as i32) - 1);
+                let window_x = (i32::from(w_x) + (x as i32)).clamp(0, (width as i32) - 1);
+
+                let (window_y, window_x) = (window_y as u32, window_x as u32);
+
+                debug_assert!(image.in_bounds(window_x, window_y));
+                // Safety: we clamped `window_x` and `window_y` to be in bounds.
+                let window_pixel = unsafe { image.unsafe_get_pixel(window_x, window_y) };
+
+                let spatial_distance = spatial_distance_lookup
+                    [(window_len * (w_y + radius) + (w_x + radius)) as usize];
+                let color_distance = color_distance.color_distance(&center_pixel, &window_pixel);
+                let weight = spatial_distance * color_distance;
+
+                (weight, window_pixel)
+            });
+
+        weighted_average(weights_and_values)
+    };
+
+    Image::from_fn(width, height, bilateral_pixel_filter)
+}
+
+fn weighted_average<P>(weights_and_values: impl Iterator<Item = (f32, P)>) -> P
+where
+    P: Pixel,
+    <P as image::Pixel>::Subpixel: 'static,
+    f32: From<P::Subpixel> + AsPrimitive<P::Subpixel>,
+{
+    let (weights_sum, weighted_channel_sums) = weights_and_values
+        .map(|(w, v)| {
+            (
+                w,
+                v.channels().iter().map(|s| w * f32::from(*s)).collect_vec(),
+            )
+        })
+        .reduce(|(w1, channels1), (w2, channels2)| {
+            (
+                w1 + w2,
+                channels1
+                    .into_iter()
+                    .zip_eq(channels2)
+                    .map(|(c1, c2)| c1 + c2)
+                    .collect_vec(),
+            )
+        })
+        .expect("cannot find a weighted average given no weights and values");
+
+    let channel_averages = weighted_channel_sums.iter().map(|x| x / weights_sum);
+
+    *P::from_slice(
+        &channel_averages
+            .map(<f32 as AsPrimitive<P::Subpixel>>::as_)
+            .collect_vec(),
+    )
+}
+
+/// Un-normalized Gaussian Weight
+fn gaussian_weight(x: f32, sigma_squared: f32) -> f32 {
+    (-0.5 * x.powi(2) / sigma_squared).exp()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_bilateral_filter_greyscale() {
```

Comment:
Let’s leave this test input and expectations the same as it was before as a sanity check - a new implementation shouldn’t change the output values enough to make this fail

Reply:
I've left the input image and input parameters the same or as their equivalents since switching from `window_size` to `radius`, hence 3 -> 1, but I think rounding differences in the new implementation are causing the output to change slightly. From my testing it still looks to be giving the same output as shown in the images I posted so I'm fairly confident it's doing the right thing.